src/main/java/org/variantsync/diffdetective/variationtree/VariationNode.java

```diff
@@ -0,0 +1,545 @@
+        if (getParent() != null) {
+            getParent().removeChild(this.cast());
+        }
+    }
+
+    /**
+     * Removes all children from the given node and adds them as children to this node.
+     * The order of the children is preserved. The given node will have no children afterwards.
+     *
+     * @param other The node whose children should be stolen.
+     * @see addChildren
+     * @see removeAllChildren
+     */
+    public void stealChildrenOf(final T other) {
+        addChildren(other.getChildren());
+        other.removeAllChildren();
+    }
+
+    /**
+     * Returns the formula that is stored in this node.
+     * The formula is not {@code null} for
+     * {@link NodeType#isConditionalAnnotation mapping nodes with annotations} and {@code null}
+     * otherwise ({@link NodeType#ARTIFACT}, {@link NodeType#ELSE}).
+     *
+     * <p>If the type parameter {@code T} of this class is not a concrete variation tree, then the
+     * returned {@link Node formula} should be treated as unmodifiable to prevent undesired side
+     * effects (e.g., to {@link DiffNode}s).
+     */
+    public abstract Node getDirectFeatureMapping();
+
+    /**
+     * Same as {@link getFeatureMapping} but returns a list of formulas representing a conjunction.
+     */
+    private List<Node> getFeatureMappingClauses() {
+        final var parent = getParent();
+
+        if (isElse() || isElif()) {
+            List<Node> and = new ArrayList<>();
+
+            if (isElif()) {
+                and.add(getDirectFeatureMapping());
+            }
+
+            // Negate all previous cases
+            var ancestor = parent;
+            while (!ancestor.isIf()) {
+                if (ancestor.isElif()) {
+                    and.add(negate(ancestor.getDirectFeatureMapping()));
+                } else {
+                    throw new RuntimeException("Expected If or Elif above Else or Elif but got " + ancestor.getNodeType() + " from " + ancestor);
+                    // Assert.assertTrue(ancestor.isArtifact());
+                }
+                ancestor = ancestor.getParent();
+            }
+            and.add(negate(ancestor.getDirectFeatureMapping()));
+
+            return and;
+        } else if (isArtifact()) {
+            return ((VariationNode<T>) parent).getFeatureMappingClauses();
+        }
+
+        return List.of(getDirectFeatureMapping());
+    }
+
+    /**
+     * Returns the full feature mapping formula of this node.
+     *
+     * <p>The feature mapping of an {@link NodeType#IF} node is its
+     * {@link DiffNode#getDirectFeatureMapping direct feature mapping}. The feature mapping of
+     * {@link NodeType#ELSE} and {@link NodeType#ELIF} nodes is determined by all formulas in the
+     * respective if-elif-else chain. The feature mapping of an {@link NodeType#ARTIFACT artifact}
+     * node is the feature mapping of its parent.
+     * See Equation (1) in <a href="https://doi.org/10.1145/3540250.3549108">our paper</a>.
```

Comment:
Use link to preprint for open access https://github.com/SoftVarE-Group/Papers/blob/raw/2022/2022-ESECFSE-Bittner.pdf